### PR TITLE
Allow setting a custom logging filter on logging handlers

### LINKED ISSUE #25981: Custom logging filters
### Discussed in https://github.com/quarkusio/quarkus/discussions/25912

<div type='discussions-op-text'>

<sup>Originally posted by **janinko** June  1, 2022</sup>
Hi,

Is there a way how to configure custom logging filter for handlers in quarkus? Something like this in WildFly: https://www.wildfly.org/news/2020/01/04/Custom-Logging-Filter/

Alternatively, is there a way how to filter logs messages that go to handled by specific Marker or logger name?</div>

### PR DESCRIPTION
**Pull request by geoand**
This PR does **not**  introduce support for logging filters on categories.

Resolves: #25981

### CODE REVIEW

File: core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingResourceProcessor.java [hunk 1]
```diff
@@ -114,6 +126,13 @@ public final class LoggingResourceProcessor {
             "isMinLevelEnabled",
             boolean.class, int.class, String.class);
 
+    private static final DotName LOGGING_FILTER = DotName.createSimple(LoggingFilter.class.getName());
+    private static final DotName FILTER = DotName.createSimple(Filter.class.getName());
+    private static final String ILLEGAL_LOGGING_FILTER_USE_MESSAGE = "'" + LoggingFilter.class.getName()
```
Comment (gastaldi) [inline core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingResourceProcessor.java:131]:
```suggestion
    private static final String ILLEGAL_LOGGING_FILTER_USE_MESSAGE = "'@" + LoggingFilter.class.getName()
```

File: core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ */
+
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface LoggingFilter {
+
+    /**
+     * Name with which the filter is referred to in configuration
+     */
+    String name();
```
Comment (gastaldi) [inline core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java:22]: Should this have a default value? The FQN perhaps?

File: core/runtime/src/main/java/io/quarkus/runtime/logging/ConsoleConfig.java [hunk 3]
```diff
@@ -55,6 +55,12 @@ public class ConsoleConfig {
     @ConfigItem(defaultValue = "0")
     int darken;
 
+    /**
+     * The names of the filter to link to the console handler.
```
Comment (gastaldi) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/ConsoleConfig.java:59]: `names` or `name`?

File: core/runtime/src/main/java/io/quarkus/runtime/logging/SyslogConfig.java [hunk 4]
```diff
@@ -89,6 +89,12 @@ public class SyslogConfig {
     @ConfigItem(defaultValue = "ALL")
     Level level;
 
+    /**
+     * The names of the filter to link to the file handler.
```
Comment (gastaldi) [inline core/runtime/src/main/java/io/quarkus/runtime/logging/SyslogConfig.java:93]: `names` or `name`?

File: core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Makes the filter class known to Quarkus by the specified name.
+ * The filter can then be configured for a handler (like the logging handler using {@code quarkus.log.console.filter}).
+ *
+ * This class must ONLY be placed on implementations of {@link java.util.logging.Filter} that are marked as {@code final}.
```
Comment (gastaldi) [inline core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java:12]: Out of curiosity: Why the `Filter` implementation class must be final?

File: core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java (hunk 2 above)
Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java:22]: I believe it's best to not have a value, that way it's very clear what the user has to configure in `application.properties`. At least that's the intention :)
If it does not work out like that in practice, we can always relax

File: core/runtime/src/main/java/io/quarkus/runtime/logging/ConsoleConfig.java (hunk 3 above)
Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/runtime/logging/ConsoleConfig.java]: Good catch, fixed

File: core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java (hunk 5 above)
Comment (geoand) [reply core/runtime/src/main/java/io/quarkus/logging/LoggingFilter.java:12]: It just makes it easier to ensure that there is ambiguity if subclasses are involved. We don't strictly need it, we can certainly relax this if people find themselves needing to subclass

File: extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java [hunk 6]
```diff
@@ -197,7 +198,7 @@ private DevServicesResultBuildItem.RunningDevService startElasticsearch(
         // Starting the server
         final Supplier<DevServicesResultBuildItem.RunningDevService> defaultElasticsearchSupplier = () -> {
             ElasticsearchContainer container = new ElasticsearchContainer(
-                    DockerImageName.parse(config.imageName));
+                    DockerImageName.parse(config.imageName).asCompatibleSubstituteFor(DEFAULT_IMAGE));
```
[the remark this answers is not in the document]
Comment (geoand) [reply extensions/elasticsearch-rest-client-common/deployment/src/main/java/io/quarkus/elasticsearch/restclient/common/deployment/DevServicesElasticsearchProcessor.java]: Oops!

**Comment by geoand** [comment]
Issues addressed

**Review by gastaldi — APPROVED** [review]

**Review by famod — APPROVED** [review]
LGTM (though I've scanned the code only briefly).

> This PR does **not** introduce support for logging filters on categories.

This would be my primary usage scenario because I usually want to filter a nagging warning for all handlers (I actually only have one most of the time), without imposing too much overhead by having the filter method called for messages from _all_ categories.
Anyway, there is nothing really preventing us from adding this in a subsequent PR, right?

File: core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingResourceProcessor.java [hunk 7]
```diff
@@ -276,6 +306,57 @@ public void run() {
...
+                throw new RuntimeException(
+                        ILLEGAL_LOGGING_FILTER_USE_MESSAGE + " Offending class is '" + classInfo.name() + "'");
+            }
+
+            MethodInfo ctor = classInfo.method("<init>");
+            if ((ctor == null) || (ctor.typeParameters().size() > 0)) {
+                throw new RuntimeException("Classes annotated with '" + LoggingFilter.class.getName()
+                        + "' must have a no-args constructor. Offending class is '" + classInfo.name() + "'");
+            }
+            filtersMap.put(instance.value("name").asString(), classInfo.name().toString());
```
[the remark this answers is not in the document]
Comment (geoand) [reply core/deployment/src/main/java/io/quarkus/deployment/logging/LoggingResourceProcessor.java:351]: I actually had that in mind but then forgot it.... Thanks for raising it!

**Comment by geoand** [comment]
> Anyway, there is nothing really preventing us from adding this in a subsequent PR, right?

We do have to be a little careful about how we handle it, as you could get unexpected results if `useParentHandlers` is set (which it is by default).

**Pull request merged**